parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt

```diff
@@ -84,28 +112,58 @@ private fun DexFile.getSuperclassName(classDefItem: ClassDefItem): String {
 /**
  * Creates new TestMethod objects with the class name changed from the super class to the subclass
  */
-private fun createTestMethodsFromSuperMethods(dexFile: DexFile, classDefItem: ClassDefItem, className: String, superTests: Set<TestMethod>): Set<TestMethod> {
-    val directory = dexFile.getAnnotationsDirectory(classDefItem)
-    val childClassAnnotations = dexFile.getClassAnnotationValues(directory)
-    val childClassAnnotationNames = childClassAnnotations.map { it.name }
-
-    val tests = superTests.map { method ->
-        val onlyParentAnnotations = method.annotations.filterNot { childClassAnnotationNames.contains(it.name) }
-                .filter { it.inherited }
-        TestMethod(className + (method.testName.substring(method.testName.indexOf('#') + 1)), onlyParentAnnotations + childClassAnnotations)
-    }
-    return tests.toSet()
-}
+private fun createTestMethodsFromSuperMethods(parsingResult: ClassParsingResult,
+                                              classTestMethods: Map<String, ClassParsingResult>,
+                                              classAllTestMethods: MutableMap<String, Set<TestMethod>>): Set<TestMethod> =
+        classAllTestMethods.getOrPut(parsingResult.className) {
+            val dexFile = parsingResult.dexFile
+
+            val superTestMethods = getSuperTestMethods(classTestMethods, parsingResult.superClassName)
+            val superSuperTestMethods = classTestMethods[parsingResult.superClassName]
+                    ?.superClassName
+                    ?.let { classTestMethods[it] }
+                    ?.let { createTestMethodsFromSuperMethods(it, classTestMethods, classAllTestMethods) }
+                    ?: emptySet()
+
+            val testMethods = superTestMethods + superSuperTestMethods
```

Comment:
Let's move recursion by one step. We can make it simpler if we work only with current and super

Reply:
sounds good 👍